aten/src/ATen/native/Distributions.cpp

```diff
@@ -198,28 +122,49 @@ Tensor _standard_gamma_grad_cpu(const Tensor& self, const Tensor& output) {
   AT_DISPATCH_FLOATING_TYPES(self.type(), "_standard_gamma_grad", [&] {
     CPU_tensor_apply3<scalar_t, scalar_t, scalar_t>(ret, self, output,
       [](scalar_t& ret_val, const scalar_t& self_val, const scalar_t &output_val) {
-         ret_val = standard_gamma_grad_one(self_val, output_val);
+        ret_val = standard_gamma_grad_one<scalar_t, scalar_t>(self_val, output_val);
       }
     );
   });
   return ret;
 }
 
-Tensor _standard_gamma_grad_cuda(const Tensor& self, const Tensor& output) {
-  AT_ERROR("_standard_gamma_grad is not implemented for CUDA types");
-}
+/*
+ * This section is a counterpart to Distributions.cu
+ */
 
 Tensor _s_poisson_cpu(const Tensor& lambda, Generator *gen) {
   Tensor ret = at::zeros(lambda.type(), lambda.sizes());
-  auto lambda_ = lambda.toType(ScalarType::Double);
   AT_DISPATCH_FLOATING_TYPES(ret.type(), "poisson", [&] {
     THGenerator* generator = get_generator(gen);
-    CPU_tensor_apply2<scalar_t, double>(ret, lambda_,
-      [generator](scalar_t& ret_val, const double& lambda){
-        ret_val = sample_poisson(lambda, generator);
+    CPU_tensor_apply2<scalar_t, scalar_t>(ret, lambda,
+      [generator](scalar_t& ret_val, const scalar_t& lambda){
+        ret_val = static_cast<scalar_t>(sample_poisson(static_cast<double>(lambda), generator));
       }
     );
-  });
+    });
+  return ret;
+}
+
+Tensor _s_gamma_cpu(const Tensor& alpha, Generator *gen) {
+  Tensor ret = alpha.type().zeros(alpha.sizes());
+  AT_DISPATCH_FLOATING_TYPES(ret.type(), "gamma", [&] {
+    THGenerator* generator = get_generator(gen);
+    CPU_tensor_apply2<scalar_t, scalar_t>(ret, alpha,
+      [generator](scalar_t& ret_val, const scalar_t& alpha){
+        BaseSampler<scalar_t> standard_uniform([generator] () {
+          return THRandom_standard_uniform(generator);
+        });
+        BaseSampler<scalar_t> standard_normal([generator] () {
+          return THRandom_normal(generator, 0.0, 1.0);
+        });
+        auto sample = sample_gamma<scalar_t, scalar_t>(alpha, standard_uniform, standard_normal);
```

Comment:
I think we should just use `double` for `accscalar_t` on CPU. There's basically no benefit of using `float` if you're not vectorizing the function, which won't happen in here.